### ISSUE TITLE
NET-1805 Fix S4201 FN: Support list pattern

### ISSUE THREAD

**Issue by cristian-ambrosini-sonarsource** [SonarSource/sonar-dotnet#6391]
### Description

FN for S4201(RedundantNullCheck).

Using C#11 list pattern, the rule is not raising an issue.

### Repro steps

```csharp
    public void SomeMethod(object[] objects)
    {
        if (objects[0] is not null && objects is [not null]) // FN
        {
        }
    }
```

### Expected behavior

The rule should raise an issue.

### Actual behavior

The rule does not raise an issue.

**Comment by martin-strecker-sonarsource** [comment]
I think this rule suffers from more severe problems than just the list pattern. This is a speculative list but from looking at the code it seems we don't support:
* `and` and `or` pattern
* conditions in other binary expression tree arms (#4645)
* implicit pattern `not null` checks (property pattern, list pattern)
* null checks with `default` instead of `null`

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1805](https://sonarsource.atlassian.net/browse/NET-1805)

[NET-1805]: https://sonarsource.atlassian.net/browse/NET-1805?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by pavel-mikula-sonarsource** [comment]
Moved
